hotham/src/systems/rendering.rs

```diff
@@ -25,72 +39,147 @@ pub fn rendering_system(
+                })
+                .instances
+                .push(Instance {
+                    transform_matrix: transform_matrix.0,
                     bounding_sphere: primitive.get_bounding_sphere(transform),
+                    skin_id,
                 });
-                draw_indirect_buffer.push(&vk::DrawIndexedIndirectCommand {
-                    index_count: primitive.indices_count,
-                    instance_count: 1,
-                    first_index: primitive.index_buffer_offset,
-                    vertex_offset: primitive.vertex_buffer_offset as _,
-                    first_instance: 0,
+        }
+    }
+
+    // Next organise this data into a layout that's easily consumed by the compute shader.
+    // ORDER IS IMPORTANT HERE! The final buffer should look something like:
+    //
+    // primitive_a
+    // primitive_a
+    // primitive_c
+    // primitive_b
+    // primitive_b
+    // primitive_e
+    // primitive_e
+    //
+    // ..etc. The most important thing is that each instances are grouped by their primitive.
+    unsafe {
+        let frame = &mut render_context.frames[render_context.frame_index];
+        let cull_data = &mut frame.primitive_cull_data_buffer;
+        cull_data.clear();
+
+        for instanced_primitive in primitive_map.values() {
+            let primitive = &instanced_primitive.primitive;
+            for instance in &instanced_primitive.instances {
+                cull_data.push(&PrimitiveCullData {
+                    draw_data: DrawData {
+                        transform: instance.transform_matrix,
+                        inverse_transpose: instance
+                            .transform_matrix
+                            .try_inverse()
+                            .unwrap()
+                            .transpose(),
+                        material_id: primitive.material_id,
+                        skin_id: instance.skin_id,
+                    },
+                    index_offset: primitive.index_buffer_offset,
+                    bounding_sphere: instance.bounding_sphere,
+                    visible: false,
                 });
             }
         }
     }
 
+    // This is the VERY LATEST we can possibly update our views, as the compute shader will need them.
     render_context.update_scene_data(views).unwrap();
+
+    // Execute the culling shader on the GPU.
     render_context.cull_objects(vulkan_context);
-    render_context.begin_pbr_render_pass(vulkan_context, swapchain_image_index);
 
-    let device = &vulkan_context.device;
-    let frame = &render_context.frames[render_context.frame_index];
-    let command_buffer = frame.command_buffer;
-    let draw_indirect_buffer = &frame.draw_indirect_buffer;
+    // Begin the render pass, bind descriptor sets.
+    render_context.begin_pbr_render_pass(vulkan_context, swapchain_image_index);
 
+    // Parse through the cull buffer and record commands. This is a bit complex.
     unsafe {
```

Comment:
It can, technically, but `frame.primitive_cull_data_buffer.as_slice()` is unsafe. Splitting it up into smaller chunks obscures the fact that we're doing *terribly unsafe* things everywhere.